cmd/microcloud/ask.go

```diff
@@ -1333,3 +1337,243 @@ func (c *initConfig) askClustered(s *service.Handler, expectedServices []types.S
+					break
+				}
+
+				// Skip systems which aren't listed in the preseed.
+				if !shared.ValueInSlice(session.Intent.Name, expectedSystems) {
+					continue
+				}
+
+				joinIntents[session.Intent.Name] = session.Intent
+				if len(joinIntents) == len(expectedSystems) {
+					renderCancel()
+				}
+
+			case <-renderCtx.Done():
+				return
+			}
+		}
+	}
+
+	var systems []types.SessionJoinPost
+	if !c.autoSetup {
+		go renderIntentsInteractive()
+
+		// Wait until the table got rendered.
+		// This is important otherwise the table might not be selectable
+		// as it's being built in a go routine.
+		select {
+		case <-rendered:
+		case <-gw.Context().Done():
+			return nil, fmt.Errorf("Failed to render join intents: %w", context.Cause(gw.Context()))
+		}
+
+		var answers []string
+		retry := false
+		err := c.askRetry("Retry selecting systems?", func() error {
+			defer func() {
+				retry = true
+			}()
+
+			fmt.Println("Select which systems you want to join:")
+
+			if retry {
+				err := table.Render(table.rows)
+				if err != nil {
+					return fmt.Errorf("Failed to render table: %w", err)
+				}
+			}
+
+			var err error
+			answers, err = table.GetSelections()
+			if err != nil {
+				return fmt.Errorf("Failed to get join intent selections: %w", err)
+			}
+
+			if len(answers) == 0 {
+				return fmt.Errorf("No system selected")
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, answer := range answers {
+			name := table.SelectionValue(answer, "NAME")
+			for intentName, intent := range joinIntents {
+				if intentName == name {
+					systems = append(systems, intent)
+				}
+			}
+		}
+	} else {
+		go renderIntents()
+
+		select {
+		case <-time.After(c.lookupTimeout):
+		case <-renderCtx.Done():
+		}
+
+		for _, name := range expectedSystems {
+			_, ok := joinIntents[name]
+			if !ok {
+				return nil, fmt.Errorf("System %q hasn't reached out", name)
+			}
+		}
+
+		for _, intent := range joinIntents {
+			systems = append(systems, intent)
+		}
+	}
+
+	return systems, nil
+}
+
+func (c *initConfig) askJoinConfirmation(gw *cloudClient.WebsocketGateway) error {
+	session := types.Session{}
+	err := gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to read join confirmation: %w", err)
+	}
+
+	if !c.autoSetup {
+		certBlock, _ := pem.Decode([]byte(session.Intent.Certificate))
+		if certBlock == nil {
+			return fmt.Errorf("Invalid certificate file")
+		}
+
+		remoteCert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		fingerprint, err := c.shortFingerprint(shared.CertFingerprint(remoteCert))
+		if err != nil {
+			return fmt.Errorf("Failed to shorten fingerprint: %w", err)
+		}
+
+		fmt.Printf("\n Received confirmation from the system %q at %q using fingerprint %q\n\n", session.Intent.Name, session.Intent.Address, fingerprint)
```

Comment:
At the point that we display this fingerprint, the mTLS has already been established, so verifying that it matches the initiator's fingerprint is a bit moot, no?

Can we move this to right after the joiner supplies the password?

Reply:
We only have the other sides fingerprint in case of mDNS.
When doing unicast the information is available as soon as the initiator talks back to us by confirming the intent to join.

The other option is we retrieve it from the underlying TLS conn when first sending the join intent.